src/util/graph.h

```diff
@@ -544,6 +560,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
   return result;
 }
 
+/// Run recursive depth-limited search on the graph, starting
+/// from multiple source nodes, to find the nodes reachable within n steps.
+/// This function initialises the search.
+/// \param src The node to start the search from.
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+/// \param limit  limit on steps
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const node_indext &src,
+  std::size_t &limit,
+  bool forwards) const
+{
+  std::vector<node_indext> start_vector(1, src);
+  return depth_limited_search(start_vector, limit, forwards);
+}
+
+/// Run recursive depth-limited search on the graph, starting
+/// from multiple source nodes, to find the nodes reachable within n steps.
+/// This function initialises the search.
+/// \param src The nodes to start the search from.
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+/// \param limit  limit on steps
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
+  bool forwards) const
+{
+  std::vector<bool> visited(nodes.size(), false);
+
+  for(const auto &node : src)
+    visited[node] = true;
+
+  return depth_limited_search(src, limit, visited, forwards);
+}
+
+/// Run recursive depth-limited search on the graph, starting
+// from multiple source nodes, to find the nodes reachable within n steps
+/// \param src The nodes to start the search from.
+/// \param limit  limit on steps
+/// \param visited vector of booleans indicating whether a node has been visited
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
+  std::vector<bool> &visited,
+  bool forwards) const
+{
+  PRECONDITION(nodes.size() < std::numeric_limits<std::size_t>::max());
```

Comment:
This is a surprising condition, effectively only ruling out the case `nodes.size() == std::numeric_limits<std::size_t>::max()`. Why is that useful?

Reply:
Ah, it should be <=, and then it rules out any case where the number of nodes is bigger than `std::numeric_limits<std::size_t>::max()`, because if it is, this function won't work because i access the nodes by an index that is a std::size_t. But I guess if the graph needs more nodes than `std::numeric_limits<std::size_t>::max()` something will fail earlier than this? I couldn't find any explicit checks elsewhere in graph.h

Reply:
The type of `nodes.size()` is (a typedef resolving to) `std::size_t`, and thus this will trivially be true. I think it's safe to drop this `PRECONDITION`.